### ISSUE TITLE
How to run the train_ego_pred.py?

### ISSUE THREAD

**Issue by Flawless1202** [MoonBlvd/tad-IROS2019#1]
I find the `train_ego_pred.py` use the data from `HEVI_dataset/ego_motion`. But I don't have the HEVI Dataset. Is there any way to train without HEVI Dataset? @MoonBlvd

**Comment by MoonBlvd** [comment]
@Flawless1202 If you would like to use your own dataset, just write your own dataloader by following the ```HEVIDataset``` in ```lib/utils/fol_dataloader.py```.  If you would like to use HEVI, I think Honda is planning to release the dataset to universities and academy in the recent 2 weeks. If you just want the ego motion data from the visual odometry, please find the data I process using ORBSLAM2 for [train](https://drive.google.com/open?id=1TE-smXm4dD2QgoCQHYmzLHqSsltoIxbe) and [validation](https://drive.google.com/open?id=1Vcu6NU7PwDOPTv6RU_7AuBfj6I0rj4dR). Each pickle file is a dictionary corresponds to one car's trajectory, and it has corresponding ego motions.

**Comment by Flawless1202** [comment]
Thanks for your reply. And I met another problem. I try to use ORB_SLAM2 to generate the ego motion data from the SA Dataset. But the videos have no camera calibration and distortion parameters. So how do you generate the ego motion data with ORB_SLAM2? @MoonBlvd

**Comment by MoonBlvd** [comment]
@Flawless1202 I didn't find intrinsic matrix neither, so I just used [fx fy cx cy] = [938.0666, 843.003, 643.298, 366.1718], which was for HEV-I. This will make the visual odometry not accurate for sure.

**Comment by Flawless1202** [comment]
@MoonBlvd Thanks for your reply. And these days I try to generate the ego motion with `ORB_SLAM2`, but the sequences is to short and often initialize failed. How do you to solve the problem? And when I try to generate the tracking result, it is not always continuous. Has this problem you ever met and how do you solve it? Look forwarding to your reply.

**Comment by MoonBlvd** [comment]
@Flawless1202 Yes, the ORB_SLAM2 won't work perfectly with video. I used linear interpolation which you can find in ```lib/utils/data_prep_utils.py```

**Comment by Flawless1202** [comment]
@MoonBlvd  Thanks. I also try to use linear interpolation and I also find it in `lib/utils/data_prep_utils.py`. But I am not really sure the data format which you called 'ORB_SLAM_FORMAT'. When I run the `orb_slam2`, I got the trajectory file format is 'timestamp x y z qx qy qz w', only 8 numbers in every line and can not use your data prepare scripts. What the data format you get from `orb_slam2`?

**Comment by srikanthmalla** [comment]
> I find the `train_ego_pred.py` use the data from `HEVI_dataset/ego_motion`. But I don't have the HEVI Dataset. Is there any way to train without HEVI Dataset? @MoonBlvd

I think if you want to try on HEVI dataset, you could change the HEVIEgoDataset loader to load the same processed pkl files like HEVIDataset loader and return ego motion only (neglecting other information). 

@MoonBlvd correct me if I am wrong. It should be sufficient right for starting (even with out downloading and processing the dataset from HEVI website) ?

**Comment by SateeshKumar21** [comment]
hey @Flawless1202  did you manage to solve this problem?

**Comment by MoonBlvd** [comment]
Hey guys, I'm sorry if I have

**Comment by MoonBlvd** [comment]
Hey guys, I'm sorry if I have closed the issue by mistake.
What I got from ORB_SLAM2 was text files. Each line of the text file was a flattend Tcws matrix. What I did [here](https://github.com/MoonBlvd/tad-IROS2019/blob/68b7686c712dd7f07ec6d8ace6f8630b62089460/lib/utils/data_prep_utils.py#L104)  was just to load the lines and reshape it to a 4*4 matrix.

**Comment by Flawless1202** [comment]
@MoonBlvd Thanks a lot. I have solved the problem, but the ORB_SLAM2 works too bad on my dataset. I don't think the odometry is accurate enough for the ego motion prediction task. Do you think the odometry is necessary?

**Comment by SateeshKumar21** [comment]
@Flawless1202 could you please explain how you solved this problem. Because right now I am also stuck at the same point that I have the 'timestamp x y z qx qy qz w' format and don't know how to convert it into the format used by @MoonBlvd.

**Comment by Flawless1202** [comment]
@SateeshKumar21 I think the `load_vo` function in `lib/utils/data_prep_utils.py` doesn't work well for the text file generated by ORB_SLAM2. So I use the [formula](https://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToMatrix/index.htm) to convert the quaternion to transformation matrix by myself.

**Comment by SateeshKumar21** [comment]
@MoonBlvd could you please specifically describe how you run ORB slam. I ran Orb-slam using the mono_TUM( https://github.com/raulmur/ORB_SLAM2/tree/master/Examples/Monocular/mono_tum) executable and I receive the text file of the format 'timestamp x y z qx qy qz w' which is not compatible with the scripts you provide.

**Comment by kimimasa-tamura-TRI** [comment]
@MoonBlvd Hi auther, first of all thank you for your great work!

As asked by others, could you please let us how to get the ORB result?
- SLAM.SaveKeyFrameTrajectoryTUM() -> outputs 'timestamp x y z qx qy qz qw', so it seems not. In addition, it output only key frames, not all the frames.
- SLAM.SaveTrajectoryTUM(); -> outputs all the frames, but not 4x4 matrix. 
- Or did you add some codes at the end of System::SaveTrajectoryTUM() like below?
```
        cv::Mat Tcw_1d = Tcw.reshape(1);
        for (int i=0; i<16; i++)
        {
            f << Tcw_1d.at<float>(i) << " ";
        }
        f << endl;
```

> Hey guys, I'm sorry if I have closed the issue by mistake. What I got from ORB_SLAM2 was text files. Each line of the text file was a flattend Tcws matrix. What I did [here](https://github.com/MoonBlvd/tad-IROS2019/blob/68b7686c712dd7f07ec6d8ace6f8630b62089460/lib/utils/data_prep_utils.py#L104) was just to load the lines and reshape it to a 4*4 matrix. @SateeshKumar21 @Flawless1202

**Comment by tianchenji** [comment]
@kimimasa-tamura-TRI Do you have any updates on this issue? I'm facing the same issue here.
